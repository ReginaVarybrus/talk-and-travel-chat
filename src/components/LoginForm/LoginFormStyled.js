import styled from 'styled-components';

export const ItemWrapp = styled('div')({
  position: 'relative',
});

export const StyledLabel = styled('label')(
  ({ color = { error: '', touched: false } }) => ({
    display: 'block',

    fontSize: '12px',
    fontWeight: 600,
    marginBottom: 8,

    color:
      color.error && color.touched
        ? 'red'
        : !color.error && color.touched
          ? 'green'
          : '#111',
  })
);

export const StyledInput = styled('input')(
  ({ color = { error: '', touched: false } }) => ({
    width: '100%',

    color: '#111',
    fontSize: '14px',
    lineHeight: '18px',

    height: 48,

    borderRadius: '8px',
    border: '1px solid',

    borderColor:
      color.error && color.touched
        ? 'red'
        : !color.error && color.touched
          ? 'green'
          : 'rgba(220, 227, 229, 0.60)',

    marginBottom: 24,
    paddingLeft: 14,

    ':hover': {
      border: '1px solid #111',
    },

    '::placeholder': {
      color: '#DCE3E5',
    },
  })
);

export const Error = styled('p')({
  position: 'absolute',
  top: 72,
  left: 10,

  color: '#DA1414',
  fontSize: '12px',
  lineHeight: '14px',
  margin: 0,
  padding: 0,
});

export const Success = styled('p')({
  position: 'absolute',
  top: 72,
  left: 10,

  color: 'green',
  fontSize: '12px',
  lineHeight: '14px',
  margin: 0,
  padding: 0,
});

export const ValidationIcon = styled('img')({
  position: 'absolute',
  top: 34,
  right: 20,
});

export const LogInBtn = styled('button')({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',

  fontSize: '14px',
  fontWeight: 600,
  lineHeight: '18px',
  letterSpacing: '-0.28px',

  width: '100%',
  height: 46,

  border: 'none',
  borderRadius: '16px',
  outline: 'none',

  background: '#3E85F3',
  boxShadow: ' 4px 2px 16px 0px rgba(136, 165, 191, 0.48)',

  cursor: 'pointer',

  color: '#ffffff',

  ':hover': {
    background: '#2B78EF',
  },
  ':focus': {
    background: '#2B78EF',
  },
  '@media (min-width: 768px)': {
    fontSize: '18px',
    lineHeight: '24px',
    letterSpacing: '-0.36px',
  },
});
