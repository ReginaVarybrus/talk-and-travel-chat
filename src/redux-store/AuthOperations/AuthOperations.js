import axios from 'axios';
import { createAsyncThunk } from '@reduxjs/toolkit';
import swal from 'sweetalert';

axios.defaults.baseURL = 'https://talk-and-travel.pp.ua';

export const token = {
  set(token) {
    axios.defaults.headers.common.Authorization = `Bearer ${token}`;
  },
  unset() {
    axios.defaults.headers.common.Authorization = '';
  },
};

export const register = createAsyncThunk('auth/register', async userData => {
  try {
    const response = await axios.post(
      '/api/authentication/register',
      userData,
      {
        headers: {
          'Content-Type': 'application/json',
        },
      }
    );

    token.set(response.data.token);
    console.log(response);
    console.log('data', response.data);

    swal('Success!', 'Letter with verification sent on your email', 'success');

    return response.data;
  } catch (e) {
    console.log(e.response.data);
    if (
      e.response.status === 400 ||
      e.response.status === 401 ||
      e.response.status === 409
    ) {
      throw new Error(swal('Error!', e.response.data.message, 'error'));
    }
  }
});

export const logIn = createAsyncThunk('auth/login', async userData => {
  try {
    const response = await axios.post('/api/authentication/login', userData, {
      headers: {
        'Content-Type': 'application/json',
      },
    });
    token.set(response.data.token);
    console.log(response);

    return response.data;
  } catch (e) {
    console.log(e.response.data);
    if (e.response.status === 400 || e.response.status === 401) {
      throw new Error(swal('Error!', e.response.data.message, 'error'));
    }
    if (e.response.status === 404) {
      throw new Error(swal('Error!', 'Email is wrong', 'error'));
    }
  }
});

export const logOut = createAsyncThunk('auth/logout', async () => {
  try {
    await axios.post('/api/authentication/logout');
    token.unset();
  } catch (e) {
    console.log(e.message);
  }
});

export const fetchCurrentUser = createAsyncThunk(
  'auth/refresh',
  async (_, thunkAPI) => {
    const state = thunkAPI.getState();
    const persistedToken = state.auth.token;

    if (persistedToken === null) {
      return thunkAPI.rejectWithValue('No valid token');
    }

    try {
      token.set(persistedToken);
      const user = await axios.get(`auth/current`);

      return user.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);

export const updateUser = createAsyncThunk(
  'user/update',
  async (user, thunkAPI) => {
    try {
      const { data } = await axios.put('/api/users', user);

      return data.user;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  }
);

// country data

export const sendDataCountryToBackend = createAsyncThunk(
  'auth/sendDataCountryToBackend',
  async ({ userId, countryDto, token }, { rejectWithValue }) => {
    try {
      const response = await axios.post(
        `/api/countries/${userId}`,
        countryDto,
        {
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
          },
        }
      );

      if (!response.data) {
        throw rejectWithValue('Response data is missing');
      }

      console.log('token', token);
      console.log('data country', response);
      return response.data;
    } catch (e) {
      if (e.response && e.response.data) {
        console.log('Error response data:', e.response.data);
        throw rejectWithValue(e.response.data.message);
      } else {
        console.error(
          'Error response is missing or does not have data property:',
          e.response
        );
        throw rejectWithValue('Unknown error occurred');
      }
    }
  }
);
