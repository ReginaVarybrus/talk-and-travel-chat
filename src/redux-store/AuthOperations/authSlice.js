import { createSlice } from '@reduxjs/toolkit';
import {
  register,
  logIn,
  logOut,
  fetchCurrentUser,
  updateUser,
  sendDataCountryToBackend,
} from './AuthOperations';

const initialState = {
  token: null,
  userDto: {
    userName: '',
    userEmail: '',
    avatar: '',
  },
  name: '',
  flagCode: '',
  userId: '',
  countryRooms: [],

  isLoggedIn: false,
  isRefresh: true,
  error: null,
};

const handlePending = state => ({
  ...state,
  isRefresh: true,
});

const handleRejected = (state, action) => ({
  ...state,
  isRefresh: false,
  error: action.payload,
});

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  extraReducers: builder =>
    builder
      .addCase(register.pending, handlePending)
      .addCase(register.rejected, handleRejected)
      .addCase(register.fulfilled, (state, action) => ({
        ...state,
        userDto: action.payload.userDto,
        token: action.payload.token,
        isLoggedIn: true,
      }))

      .addCase(logIn.pending, handlePending)
      .addCase(logIn.rejected, handleRejected)
      .addCase(logIn.fulfilled, (state, action) => ({
        ...state,
        userDto: action.payload.userDto,
        token: action.payload.token,
        isLoggedIn: true,
        error: null,
      }))

      .addCase(logOut.pending, handlePending)
      .addCase(logOut.rejected, handleRejected)
      .addCase(logOut.fulfilled, state => ({
        ...state,
        userDto: { userName: null, userEmail: null },
        token: null,
        isLoggedIn: false,
        isRefresh: false,
      }))

      .addCase(fetchCurrentUser.pending, handlePending)
      .addCase(fetchCurrentUser.rejected, handleRejected)
      .addCase(fetchCurrentUser.fulfilled, (state, action) => ({
        ...state,
        userDto: action.payload.userDto,
        isLoggedIn: true,
        isRefresh: false,
      }))

      .addCase(updateUser.fulfilled, (state, action) => ({
        ...state,
        userDto: { ...state.userDto, ...action.payload },
      }))
      .addCase(sendDataCountryToBackend.pending, handlePending)
      .addCase(sendDataCountryToBackend.rejected, (state, action) => {
        handleRejected(state, action);
      })
      .addCase(sendDataCountryToBackend.fulfilled, (state, action) => ({
        ...state,
        name: action.payload.name,
        flagCode: action.payload.flagCode,
        isLoggedIn: true,
      })),
});

export default authSlice.reducer;
