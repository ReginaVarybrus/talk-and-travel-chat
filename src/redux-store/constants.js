const ULRs = {
  register: 'authentication/register',
  login: 'authentication/login',
  logout: 'authentication/logout',
  currentUser: 'current',
  updateUser: 'users',
  userInfo: userId => `users/${userId}`,
  getMainCountryChatByName: countryName => `chats/${countryName}/main`,
  userCountries: userId => `countries/user/${userId}`,
  subscriptionToMessages: chatId => `/countries/${chatId}/messages`,
  subscriptionToUserErrors: userId => `/user/${userId}/errors`,
  getChatsParticipants: chatId => `/chats/${chatId}/users`,
  joinToGroupChat: '/chat/events.joinChat',
  leaveOutGroupChat: '/chat/events.leaveChat',
  startTyping: '/chat/events.startTyping',
  stopTyping: '/chat/events.stopTyping',
  getPrivateChats: userId => `chats/user/${userId}/private`,
  createPrivateChat: 'chats/private',
  getChatsMessages: chatId => `chats/${chatId}`,
};

export default ULRs;
