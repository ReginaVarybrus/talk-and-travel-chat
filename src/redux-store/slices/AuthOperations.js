/* eslint-disable consistent-return */
import { createAsyncThunk } from '@reduxjs/toolkit';
import swal from 'sweetalert';

import { token as authToken, axiosClient } from '@/services/api';
import { clearUser, setUsers } from '@/redux-store/slices/userSlice';
import URLs from '@/constants/constants';

export const register = createAsyncThunk(
  'auth/register',
  async (userData, { rejectWithValue }) => {
    try {
      const response = await axiosClient.post(URLs.register, userData);

      swal(
        'Success!',
        'Please check your email for a confirmation link to complete the registration.',
        'success'
      );
      return response.data;
    } catch (e) {
      swal('Error!', e.response.data.message || 'Registration failed', 'error');
      return rejectWithValue(e.response.data.message);
    }
  }
);

export const logIn = createAsyncThunk(
  'auth/login',
  async (userData, { dispatch, rejectWithValue }) => {
    try {
      const response = await axiosClient.post(URLs.login, userData);
      dispatch(setUsers(response.data.userDto));
      authToken.set(response.data.token);

      return response.data;
    } catch (e) {
      if (e.response.status === 400 || e.response.status === 401) {
        swal('Error!', 'Email or password invalid', 'error');
      } else if (e.response.status === 404) {
        swal('Error!', 'Email is wrong', 'error');
      } else {
        swal('Error!', 'Login failed', 'error');
      }
      return rejectWithValue(e.response.data.message);
    }
  }
);

export const logOut = createAsyncThunk(
  'auth/logOut',
  async (_, { dispatch }) => {
    try {
      await axiosClient.post(URLs.logout);
      authToken.unset();
      dispatch(clearUser());
    } catch (error) {
      throw new Error(error.message);
    }
  }
);

export const verifyEmail = createAsyncThunk(
  'auth/verifyEmail',
  async (verificationToken, { dispatch, rejectWithValue }) => {
    try {
      const response = await axiosClient.post(URLs.verifyEmail, {
        token: verificationToken,
      });
      const { token, userDto } = response.data;
      authToken.set(token);
      dispatch(setUsers(userDto));

      swal(
        'Success!',
        'Your email has been confirmed. Welcome to the app!',
        'success'
      );
      return response.data;
    } catch (error) {
      swal('Error!', 'Verification failed. Please try again.', 'error');
      return rejectWithValue(error.response.data.message);
    }
  }
);

export const logInWithGoogle = createAsyncThunk(
  'auth/logInWithGoogle',
  async (googleData, { dispatch, rejectWithValue }) => {
    try {
      console.log('Attempting login with Google data:', googleData);
      const response = await axiosClient.post(URLs.loginWithSocial, {
        userEmail: googleData.email,
      });

      console.log('Login successful, server response:', response.data);
      dispatch(setUsers(response.data.userDto));

      return response.data;
    } catch (err) {
      console.error('Login failed, error response:', err.response);
      if (err.response?.status === 401 || err.response?.status === 404) {
        console.log('User not found, attempting registration...');
        await axiosClient.post(URLs.registerWithSocial, {
          userName: googleData.name,
          userEmail: googleData.email,
        });
        console.log('Registration successful, retrying login...');
        const response = await axiosClient.post(URLs.loginWithSocial, {
          userEmail: googleData.email,
        });

        console.log('Login after registration successful:', response.data);
        dispatch(setUsers(response.data.userDto));

        return response.data;
      }
      return rejectWithValue(err.response?.data.message || 'Login failed');
    }
  }
);

export const logInWithFacebook = createAsyncThunk(
  'auth/logInWithFacebook',
  async (facebookData, { rejectWithValue }) => {
    console.log('Facebook Data:', facebookData);
    try {
      const response = await axiosClient.post(URLs.loginWithSocial, {
        userEmail: facebookData.email,
      });
      console.log('Login response:', response.data);
      return response.data;
    } catch (err) {
      console.error('Login error:', err);
      if (err.response?.status === 404) {
        console.log('User not found, registering...');
        await axiosClient.post(URLs.registerWithSocial, {
          userName: facebookData.name,
          userEmail: facebookData.email,
        });
        const retryResponse = await axiosClient.post(URLs.loginWithSocial, {
          userEmail: facebookData.email,
        });
        console.log('Retry login response:', retryResponse.data);
        return retryResponse.data;
      }
      return rejectWithValue(err.response?.data.message || 'Login failed');
    }
  }
);
