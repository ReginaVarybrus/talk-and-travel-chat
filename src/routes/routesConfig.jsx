import { lazy } from 'react';
import PrivateRoute from '@/routes/PrivateRoute';
import { createBrowserRouter } from 'react-router-dom';

const ROOT = import.meta.env.BASE_URL;

export const routesPath = {
  MAIN: ROOT,
  LOGIN: `${ROOT}login`,
  REGISTER: `${ROOT}register`,
  APP: `${ROOT}app`,
  CHAT: `${ROOT}app/chat/`,
  ACCOUNT: `${ROOT}app/account/`,
};

const importComponent = {
  MAIN: lazy(() => import('@/routes/MainRoute/MainRoute')),
  LOGIN: lazy(() => import('@/routes/LoginRoute/LoginRoute')),
  LAYOUT: lazy(() => import('@/components/Layout/Layout')),
  REGISTER: lazy(() => import('@/routes/RegisterRoute/RegisterRoute')),
  CHAT: lazy(() => import('@/routes/ChatRoute/ChatRoute')),
  ACCOUNT: lazy(() => import('@/routes/AccountRoute/AccountRoute')),
  ERROR_COMPONENT: lazy(() => import('@/routes/Page404/Page404')),
};

export const router = createBrowserRouter([
  {
    path: routesPath.MAIN,
    Component: importComponent.MAIN,
    errorElement: importComponent.ERROR_COMPONENT,
  },
  {
    path: routesPath.LOGIN,
    Component: importComponent.LOGIN,
  },
  {
    path: routesPath.REGISTER,
    Component: importComponent.REGISTER,
  },
  {
    path: routesPath.APP,
    element: (
      <PrivateRoute
        component={importComponent.LAYOUT}
        redirectTo={routesPath.LOGIN}
      />
    ),
    children: [
      {
        path: routesPath.CHAT,
        element: (
          <PrivateRoute
            component={importComponent.CHAT}
            redirectTo={routesPath.LOGIN}
          />
        ),
      },
      {
        path: routesPath.ACCOUNT,
        element: (
          <PrivateRoute
            component={importComponent.ACCOUNT}
            redirectTo={routesPath.LOGIN}
          />
        ),
      },
    ],
  },
]);
