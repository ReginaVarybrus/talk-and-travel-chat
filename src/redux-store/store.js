import { configureStore, combineSlices } from '@reduxjs/toolkit';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { authSlice } from './slices/authSlice';
import { userSlice } from './slices/userSlice';

const authPersistConfig = {
  key: 'auth',
  storage,
  whitelist: ['token'],
};

const rootPersistConfig = {
  key: 'root',
  storage,
  blacklist: ['auth'],
};

const rootReducer = combineSlices({
  auth: persistReducer(authPersistConfig, authSlice.reducer),
  user: userSlice.reducer,
});

const persistedReducer = persistReducer(rootPersistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);
