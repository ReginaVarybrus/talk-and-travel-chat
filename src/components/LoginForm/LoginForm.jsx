import { useFormik } from 'formik';
import { useNavigate } from 'react-router-dom';
import { logIn } from '@/redux-store/AuthOperations/AuthOperations';
import { useDispatch } from 'react-redux';
import { routesPath } from '@/routes/routesConfig';

import {
  formFields,
  schema,
} from '@/components/LoginForm/LoginValidationSchema';
import InputField from '@/components/InputField/InputField';

import {
  ButtonBlock,
  RegisterFormBackground as LoginFormBackground,
  RegisterFormContainer as LoginFormContainer,
  SignInFormStyles as LoginFormStyles,
  SignInTitle as LoginTitle,
  SignInText as LoginText,
  LoginLink as SignUpLink,
  SignUpBtn as LogInBtn,
  Separator,
} from '@/components/RegisterForm/RegisterForm.styled';

import ButtonFacebook from '@/components/Buttons/FaceBook/FaceBookButton';
import ButtonGoogle from '@/components/Buttons/GoogleButton/GoogleButton';

const initialValues = {};
Object.keys(formFields).forEach(key => {
  initialValues[key] = '';
});

const LoginForm = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const redirect = () => {
    navigate(routesPath.REGISTER);
  };

  const formik = useFormik({
    initialValues,
    validationSchema: schema,
    validateOnChange: false,
    onSubmit: (values, { resetForm }) => {
      dispatch(logIn(values));
      navigate(routesPath.ROOMS);
      resetForm();
    },
  });

  return (
    <LoginFormBackground>
      <LoginFormContainer>
        <LoginFormStyles onSubmit={formik.handleSubmit} autoComplete="off">
          <LoginTitle>Welcome back</LoginTitle>
          <LoginText>
            Don`t have an account yet?{' '}
            <SignUpLink onClick={redirect}>Sign up</SignUpLink>
          </LoginText>
          {Object.entries(formFields).map(([key, value]) => (
            <InputField
              key={key}
              props={value}
              formik={formik}
              name={value.general}
            />
          ))}
          <LogInBtn type="submit">Log In</LogInBtn>
          <Separator />
          <ButtonBlock>
            <ButtonGoogle type="button" />
            <ButtonFacebook type="button" />
          </ButtonBlock>
        </LoginFormStyles>
      </LoginFormContainer>
    </LoginFormBackground>
  );
};

export default LoginForm;
