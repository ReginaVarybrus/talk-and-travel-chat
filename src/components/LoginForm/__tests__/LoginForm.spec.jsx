import { description, expect, it, vi, mock } from 'vitest';
import { render, screen, fireEvent, within } from '@testing-library/react';

import { Provider } from 'react-redux';
import { RouterProvider } from 'react-router-dom';
import { persistor, store } from '@/redux-store/store';
import { PersistGate } from 'redux-persist/integration/react';
import { router } from '@/routes/routesConfig';

import { useFormik } from 'formik';
import { useNavigate } from 'react-router-dom';
import { logIn } from '@/redux-store/AuthOperations/AuthOperations';

import LoginForm from '../LoginForm';

const useFormikMock = vi.fn(() => ({
  initialValues: { userEmail: '', password: '' },
  validationSchema: {},
  validateOnChange: false,
  onSubmit: vi.fn(),
}));
const mockDispatch = vi.fn();
const mockResetForm = vi.fn();

const useNavigateMock = vi.fn();
const logInMock = vi.fn();

vi.mock('formik', () => ({
  useFormik: vi.fn(),
}));

vi.mock('react-router-dom', () => ({
  ...require('react-router-dom'),
  navigate: vi.fn(),
}));

vi.mock('@/redux-store/AuthOperations/AuthOperations', async importOriginal => {
  const actual = await importOriginal();
  return {
    ...actual,
    logIn: vi.fn(),
  };
});

describe('LoginForm component test', () => {
  it('should render the LoginForm component', () => {
    useFormik.mockReturnValue({
      ...useFormikMock(),
      onSubmit: useFormikMock().onSubmit,
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <RouterProvider router={router}>
          <LoginForm />
        </RouterProvider>
      </Provider>
    );

    const userEmailInput = getByTestId('userEmail');
    fireEvent.change(userEmailInput, { target: { value: 'test@example.com' } });

    const passwordInput = getByTestId('password');
    fireEvent.change(passwordInput, { target: { value: 'password123' } });

    fireEvent.submit(getByTestId('loginForm'));

    expect(useFormikMock().onSubmit).toHaveBeenCalledWith(
      { userEmail: 'test@example.com', password: 'password123' },
      { resetForm: mockResetForm }
    );

    expect(navigate).toHaveBeenCalledWith(routesPath.CHAT);
  });

  it('displays error for invalid email', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <RouterProvider router={router}>
          <LoginForm />
        </RouterProvider>
      </Provider>
    );
    const emailInput = getByTestId('userEmail');
    fireEvent.change(emailInput, { target: { value: 'invalid-email' } });

    const error = screen.getByText('Invalid email address');
    expect(error).toBeInTheDocument();
  });
});
