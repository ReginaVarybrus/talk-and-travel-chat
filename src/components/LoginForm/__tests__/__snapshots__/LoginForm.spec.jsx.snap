// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm component test > should render the LoginForm component 1`] = `"<div></div>"`;
